drivers/usb/bc12/bc12_pi3usb9201.c

```diff
@@ -0,0 +1,371 @@
+	pi3usb9201_interrupt_enable(dev, false);
+
+	/*
+	 * Let the application know there's no more charge available for the
+	 * supplier type that was most recently detected.
+	 */
+	pi3usb9201_update_charging_partner(dev, NULL);
+
+	return 0;
+}
+
+static int pi3usb9201_set_portable_device(const struct device *dev)
+{
+	/* Disable interrupts during mode change */
+	pi3usb9201_interrupt_enable(dev, false);
+
+	if (pi3usb9201_client_detect_start(dev)) {
+		struct bc12_partner_state new_result;
+
+		/*
+		 * VBUS is present, but starting bc1.2 detection failed
+		 * for some reason. Set the partner type to unknown to disable
+		 * USB data signals and limit current draw to 100 mA.
+		 */
+		new_result.voltage_mv = BC12_CHARGER_VOLTAGE_MV;
+		new_result.current_ma = BC12_CHARGER_MIN_CURR_MA;
+		new_result.type = BC12_TYPE_UNKNOWN;
+		/* Save supplier type and notify callbacks */
+		pi3usb9201_update_charging_partner(dev, &new_result);
+		LOG_ERR("bc1.2 detection failed, using defaults");
+
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static void pi3usb9201_isr_work(struct k_work *item)
+{
+	struct pi3usb9201_data *pi3usb9201_data = CONTAINER_OF(item, struct pi3usb9201_data, work);
+	const struct device *dev = pi3usb9201_data->dev;
+	uint8_t client;
+	uint8_t host;
+	int rv;
+
+	rv = pi3usb9201_get_status(dev, &client, &host);
+	if (rv < 0) {
+		LOG_ERR("Failed to get host/client status");
+		return;
+	}
+
+	if (client != 0) {
+		/*
+		 * Any bit set in client status register indicates that
+		 * BC1.2 detection has completed.
+		 */
+		pi3usb9201_client_detect_finish(dev, client);
+	}
+}
+
+static void pi3usb9201_gpio_callback(const struct device *dev, struct gpio_callback *cb,
+				     uint32_t pins)
+{
+	struct pi3usb9201_data *pi3usb9201_data = CONTAINER_OF(cb, struct pi3usb9201_data, gpio_cb);
+
+	k_work_submit(&pi3usb9201_data->work);
+}
+
+static int pi3usb9201_set_role(const struct device *dev, enum bc12_role role)
+{
+	switch (role) {
+	case BC12_DISCONNECTED:
+		return pi3usb9201_disconnect(dev);
+	case BC12_PORTABLE_DEVICE:
+		return pi3usb9201_set_portable_device(dev);
+
+	default:
+		LOG_ERR("unsupported BC12 role: %d", role);
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int pi3usb9201_set_result_cb(const struct device *dev, bc12_callback_t cb, void *user_data)
+{
+	struct pi3usb9201_data *pi3usb9201_data = dev->data;
+
+	pi3usb9201_data->result_cb = cb;
+	pi3usb9201_data->result_cb_data = user_data;
+
+	return 0;
+}
+
+static const struct bc12_driver_api pi3usb9201_driver_api = {
+	.set_role = pi3usb9201_set_role,
+	.set_result_cb = pi3usb9201_set_result_cb,
+};
+
+static int pi3usb9201_init(const struct device *dev)
+{
+	const struct pi3usb9201_config *cfg = dev->config;
+	struct pi3usb9201_data *pi3usb9201_data = dev->data;
+
+	if (!i2c_is_ready_dt(&cfg->i2c)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	if (!gpio_is_ready_dt(&cfg->intb_gpio)) {
+		LOG_ERR("int_gpiob device is not ready");
```

Comment:
What is "int_gpiob device" or is it "intb_gpio" and why not "INTB GPIO device"?

Reply:
Fixed